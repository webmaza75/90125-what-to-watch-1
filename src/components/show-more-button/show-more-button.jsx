import React from 'react';
import {connect} from 'react-redux';
import {ActionCreator} from '../../actions/actions.js';
import {
  getFilmsByGenre,
  getMaxShowFilms
} from '../../reducers/films/selectors.js';
import {
  itemShape
} from '../../models.js';
import PropTypes from 'prop-types';

const ShowMoreButton = (props) => {
  const {
    onShowMoreClick,
    maxShowFilms,
    filmsGroup
  } = props;

  if (filmsGroup.length < maxShowFilms) {
    return null;
  }

  return <div className="catalog__more">
    <button className="catalog__button" type="button" onClick={
      (event) => {
        event.preventDefault();
        onShowMoreClick();
      }
    }>Show more</button>
  </div>;
};

ShowMoreButton.propTypes = {
  onShowMoreClick: PropTypes.func,
  maxShowFilms: PropTypes.number,
  filmsGroup: PropTypes.arrayOf(itemShape)
};

const mapStateToProps = (state) => ({
  maxShowFilms: getMaxShowFilms(state),
  filmsGroup: getFilmsByGenre(state)
});

export {ShowMoreButton};

export default connect(
    mapStateToProps,
    {
      onShowMoreClick: ActionCreator.increaseMaxShowFilms
    }
)(ShowMoreButton);
