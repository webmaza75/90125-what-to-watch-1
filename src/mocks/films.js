const filmList = [
  {
    backgroundColor: `#9B7E61`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    director: `Steven Spielberg`,
    genre: `Adventure`,
    id: 1,
    isFavorite: false,
    title: `War of the Worlds`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg`,
    src: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: 6.5,
    ratingLevel: `Good`,
    released: 2005,
    runTime: 116,
    scoresCount: 386834,
    starring: [
      `Tom Cruise`,
      `Dakota Fanning`,
      `Tim Robbins`,
    ],
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`
  },
  {
    backgroundColor: `#F1E9CE`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    title: `Macbeth`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 6.6,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    starring: [
      `Michael Fassbender`,
      `Marion Cotillard`,
      `Jack Madigan`
    ],
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`
  },
  {
    backgroundColor: `#D8E3E5`,
    backgroundImage: `https://es31-server.appspot.com/wtw/static/film/background/Moonrise_Kingdom.jpg`,
    description: `A pair of young lovers flee their New England town, which causes a local search party to fan out to find them.`,
    director: `Wes Anderson`,
    genre: `Adventure`,
    id: 3,
    isFavorite: false,
    title: `Moonrise Kingdom`,
    posterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Moonrise_Kingdom.jpg`,
    picture: `https://es31-server.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: 7.8,
    ratingLevel: `Good`,
    released: 2012,
    runTime: 94,
    scoresCount: 291183,
    starring: [
      `Jared Gilman`,
      `Kara Hayward`,
      `Bruce Willis`
    ],
    videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`
  }
];

export default filmList;
