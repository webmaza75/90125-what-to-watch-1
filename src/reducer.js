import ActionTypes from './action-types.js';
import filmList from './mocks/films.js';
import {ALL_GENRES} from './consts.js';

const initialState = {
  films: filmList,
  filter: ALL_GENRES,
  filmsGroup: filmList
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.CHANGE_GENRE:
      return {
        ...state,
        filter: action.payload
      };
    case ActionTypes.GET_FILMS_BY_GENRE:
      if (action.payload.filter === ALL_GENRES) {
        return initialState;
      }
      return {
        ...state,
        filmsGroup: action.payload.films.filter(({genre}) => genre.includes(action.payload.filter))
      };
  }
  return state;
};

export default reducer;
