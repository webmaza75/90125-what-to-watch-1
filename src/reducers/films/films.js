import ActionTypes from '../../actions/action-types.js';
import {ALL_GENRES} from '../../consts.js';
import {ActionCreator} from '../../actions/actions.js';
import {sortComments} from '../../utils.js';
import {getMovieRatingLevel} from '../../utils.js';

const initialState = {
  films: [],
  filter: ALL_GENRES,
  comments: []
};

export const transform = (data) => ({
  backgroundColor: data.background_color,
  backgroundImage: data.background_image,
  description: data.description,
  director: data.director,
  genre: data.genre,
  id: data.id,
  isFavorite: data.is_favorite,
  title: data.name,
  posterImage: data.poster_image,
  picture: data.preview_image,
  src: data.preview_video_link,
  rating: data.rating,
  ratingLevel: getMovieRatingLevel(data.rating),
  released: data.released,
  runTime: data.run_time,
  scoresCount: data.scores_count,
  starring: data.starring,
  videoLink: data.video_link
});

const Operation = {
  loadFilms: () => (dispatch, getState, api) => {
    return api.get(`/films`)
    .then((response) => {
      const result = response.data.map(transform);
      dispatch(ActionCreator.loadFilms(result));
    });
  },
  loadComments: (filmId) => (dispatch, getState, api) => {
    return api.get(`/comments/${filmId}`)
    .then((response) => {
      const result = response.data.sort(sortComments);
      dispatch(ActionCreator.loadComments(result));
    });
  },
  addComment: (id, params) => (dispatch, getState, api) => {
    return api
      .post(`/comments/${id}`, params)
      .then((res) => {
        dispatch(ActionCreator.addComment(res.data));
      }).catch((error) => {
        dispatch(ActionCreator.submitCommentsError(error.response.data.error));
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.LOAD_FILMS:
      return ({
        ...state,
        films: action.payload,
      });
    case ActionTypes.CHANGE_GENRE:
      return {
        ...state,
        filter: action.payload
      };
    case ActionTypes.LOAD_COMMENTS:
      return ({
        ...state,
        comments: action.payload,
      });
    case ActionTypes.ADD_COMMENT:
      return ({
        ...state,
        comments: action.payload,
      });
  }
  return state;
};

export {
  Operation,
  reducer
};
