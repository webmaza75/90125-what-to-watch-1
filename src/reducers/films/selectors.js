import {createSelector} from 'reselect';
import NameSpace from '../name-spaces.js';
import {ALL_GENRES} from '../../consts.js';

const NAME_SPACE = NameSpace.FILMS;

export const getFilms = (state) => {
  return state[NAME_SPACE].films;
};

export const getActiveFilter = (state) => {
  return state[NAME_SPACE].filter;
};

export const getFilmsByGenre = createSelector(
    getFilms,
    getActiveFilter,
    (films, filter) => {
      if (filter === ALL_GENRES) {
        return films;
      }
      return films.filter((item) => item.genre === filter);
    }
);

export const getGenres = createSelector(
    getFilms,
    (films) => {
      const allGenres = films.map(({genre}) => genre);
      return [ALL_GENRES, ...new Set(allGenres)];
    }
);

export const getFavoriteList = createSelector(
    getFilms,
    (films) => {
      return films.filter((item) => item.isFavorite === true);
    }
);

export const getMovieById = (state, filmId) => {
  return state[NAME_SPACE].films.find((item) => item.id === +filmId);
};

export const getComments = (state) => {
  return state[NAME_SPACE].comments;
};
