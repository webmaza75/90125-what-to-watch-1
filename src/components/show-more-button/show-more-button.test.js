import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {createStore} from 'redux';

import {ShowMoreButton} from './show-more-button.jsx';
import reducer from '../../reducers/reducer.js';
import films from '../../mocks/films.js';

describe(`ShowMoreButton`, () => {
  it(`ShowMoreButton correctly renders when films more when limit to show`, () => {
    const showMoreButton = renderer
      .create(<Provider store={createStore(reducer)}>
        <ShowMoreButton maxShowFilms={2} filmsGroup={films} />
      </Provider>)
      .toJSON();

    expect(showMoreButton).toMatchSnapshot();
  });

  it(`ShowMoreButton correctly renders when films count equal to limit to show`, () => {
    const showMoreButton = renderer
      .create(<Provider store={createStore(reducer)}>
        <ShowMoreButton maxShowFilms={3} filmsGroup={films} />
      </Provider>)
      .toJSON();

    expect(showMoreButton).toMatchSnapshot();
  });

  it(`ShowMoreButton doesn\`t render when films count less when limit to show`, () => {
    const showMoreButton = renderer
      .create(<Provider store={createStore(reducer)}>
        <ShowMoreButton maxShowFilms={20} filmsGroup={films} />
      </Provider>)
      .toJSON();

    expect(showMoreButton).toMatchSnapshot();
  });
});
