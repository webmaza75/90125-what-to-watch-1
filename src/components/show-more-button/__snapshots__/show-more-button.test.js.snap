// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMoreButton ShowMoreButton correctly renders when films count equal to limit to show 1`] = `
<div
  className="catalog__more"
>
  <button
    className="catalog__button"
    onClick={[Function]}
    type="button"
  >
    Show more
  </button>
</div>
`;

exports[`ShowMoreButton ShowMoreButton correctly renders when films more when limit to show 1`] = `
<div
  className="catalog__more"
>
  <button
    className="catalog__button"
    onClick={[Function]}
    type="button"
  >
    Show more
  </button>
</div>
`;

exports[`ShowMoreButton ShowMoreButton doesn\`t render when films count less when limit to show 1`] = `null`;
